### ISSUE TITLE
Verbatim block followed by markdown header breaks verbatim block

### ISSUE THREAD

**Issue by Baertierchen** [doxygen/doxygen#8308]
**Describe the bug**
A verbatim block that is immediately followed by a markdown header breaks the verbatim block and leads to a warning:

> warning: unexpected command endverbatim

Example code:
```
/*!
 \brief dummy
 \verbatim
 Foo Bar
 \endverbatim
 ---------------
 Markdown Header
 ---------------
 */
struct Foo {};
```
This code is rendered correctly by Doxygen 1.8.11.
In Doxygen 1.8.13, the markdown header is no longer generated.
As of Doxygen 1.8.16, the header is again generated, but now the problem described above occurs.

**Expected behavior**
Correctly parsed documentation with working verbatim block - no warnings (as it was in Doxygen 1.8.11).

**Screenshots**

|Expected (Doxygen 1.8.11)|Missing Header (Doxygen - 1.8.12 - 1.8.15)|Broken documentation (Doxygen 1.8.16 - 1.9.0)|
|-----|-----|------|
|![Doxy_Valid_11](https://user-images.githubusercontent.com/4766432/103648873-d50ec580-4f5d-11eb-82e7-7ffea4f4261f.png)|![Doxy_Broken_15](https://user-images.githubusercontent.com/4766432/103649168-4babc300-4f5e-11eb-8ecd-82a8dc7d312f.png)|![Doxy_Broken](https://user-images.githubusercontent.com/4766432/103647964-68df9200-4f5c-11eb-9275-4f32404a3514.png)|

**To Reproduce**
Put the code snippet above in a cpp-file and run doxygen with default configuration.

**Version**
Versions 1.8.11 (last working) - 1.9.0 (still broken)

**Comment by albert-github** [comment]
It looks like that that something is going wrong in the markdown translation as this gives (`doxygen -d markdown`):
```
---- input -------

\brief dummy
\verbatim
Foo Bar
\endverbatim
---------------
Markdown Header
---------------

---- output -----

\brief dummy
@subsection autotoc_md0 \verbatim
Foo Bar
\endverbatim

@subsection autotoc_md1 Markdown Header
```

As a workaround: you can overcome the problem by placing an empty line before the first `---------------` line.

**Comment by albert-github** [comment]
The essential remark in the documentation is (see also https://www.doxygen.nl/manual/markdown.html#md_rulers):
> Note that when using hyphens and the previous line is not empty you have to use at least one whitespace in the
sequence of hyphens otherwise it might be seen as a level 2 header

This remark was not yet present in the 1.8.11 manual though, but I think that the handling as done in 1.8.11 was a bug (and there a fortunate circumstance).
Between 1.8.11 version and the current version the markdown engine has been made better and also placed at a better place.

In this case we have before the first `---------------` line not an empty line (and no embedded spaces) so it is see as a level 2 header indicator and hence the problems.

It is not clear whether the first `---------------` is meant as a "nice" separator in human readable code or is intended as a horizontal ruler. In the first case best is to remove this line (or maybe replace it with e.g. 2 empty lines), when intended (as is also shown in the 1.8.11 version) as a horizontal ruler one should place an empty line before it (or use embedded spaces).

I think this problem is usage.

**Comment by Baertierchen** [comment]
The documentation says that if I want to place a ruler, I should make sure that the line above the ruler is empty to avoid confusion with headers. That makes perfect sense!
In my example, I want to have a header after a verbose section. So this not does not apply here since I do actually want to have this interpreted as a header.
And in any case, the current implementation breaks the verbose block above the markdown section. IMO, this is definitely unintended behavior.

**Comment by albert-github** [comment]
When you add the empty line or remove the `------------` line or adjust it with space don't you get the behavior you want?

I think the 1.8.11 version showed the unintended behavior.
So e.g.

```
/*!
 \brief dummy
 \verbatim
 Foo Bar
 \endverbatim

 Markdown Header
 ---------------
 */
struct Foo {};
```
or
```
/*!
 \brief dummy
 \verbatim
 Foo Bar
 \endverbatim
 
 ---------------
 Markdown Header
 ---------------
 */
struct Foo {};
```

**Comment by Baertierchen** [comment]
Yes, this confirms that there is a workaround - there is still no reason why the original version does not work, so that is a bug.

In my specific case, we are using file headers with a unified format - and that format contains this verbose section followed by the markdown header. Changing that header is of course not impossible, but quite some effort. So effectively, this is a blocker that prevents us from updating to a Doxygen version later than 1.8.11.

**Comment by albert-github** [comment]
Unfortunately I only see the possibility for you to update those file headers with a unified format.
How many files it would be but I think it would be quite doable (with the right editor).

**Comment by albert-github** [comment]
I still think the best solution, at the moment for sure, is to add the empty line, but the problem remains an intriguing problem.

@doxygen the commands like `\verbatim` ...`\endverbatim` (and also `code`, `*only`, `\f*` etc.) have a bit of a special meaning and maybe they should be handled specifically in `QCString Markdown::processBlocks(const QCString &s,int indent)` as is done at some other places as well.
- What is your opinion?

**Comment by doxygen** [comment]
@albert-github yes I think so, it is hard to understand for a user why the output is as it is. Basically the verbatim section is now (partly!) seen as the header title, i.e. similar to using `\section label \verbatim some text \endverbatim` which also doesn't work.

**Comment by albert-github** [comment]
I've just pushed a proposed patch, pull request #8314

**Comment by albert-github** [comment]
Code has been integrated in master on github (please don't close the issue as this will be done at the moment of an official release).

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.9.2.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
